src/wurf/existing_tag_resolver.py

```diff
@@ -2,37 +2,83 @@
 # encoding: utf-8
 
 import os
+import json
 
 class ExistingTagResolver(object):
     """
     Resolves a specific semver version if a compatible tag is already checked
     out, and there is no newer version in the tag database.
     """
 
-    def __init__(self, ctx, dependency, semver_selector, tag_database,
-                 parent_folder, sources):
+    def __init__(self, ctx, dependency, semver_selector, tag_database, resolver,
+        cwd):
         """ Construct a new ExistingTagResolver instance.
 
         :param ctx: A Waf Context instance.
         :param dependency: The Dependency object.
         :param semver_selector: A SemverSelector instance.
         :param tag_database: A TagDatabase instance.
-        :param parent_folder: A ParentFolder instance.
-        :param sources: The URLs of the dependency as a list of strings
+        :param resolver: A resolver instance.
+        :param cwd: Current working directory as a string. This is the place
+            where we should create new folders etc.
         """
         self.ctx = ctx
         self.dependency = dependency
         self.semver_selector = semver_selector
         self.tag_database = tag_database
-        self.parent_folder = parent_folder
-        self.sources = sources
+        self.resolver = resolver
+        self.cwd = cwd
 
     def resolve(self):
         """
         Resolves the path to an existing checkout folder.
 
         :return: The path to the existing tag, otherwise None.
         """
+
+        tags = self.__load_tag_file()
+
+        # Try to resolve path from tag file and tag database
+        path = self.__resolve_path(tags=tags)
+
+        if path:
+            return path
+
+        # Fallback to the resolver
+        path = self.resolver.resolve()
+
+        assert os.path.isdir(path)
```

Comment:
Just tested this with an optional dependency, the underlying resolver returned None, so this assert is not okay here.